Enable provisioning of Windows nodes using vSphere templates

Ihor-kolomiiets opened the pull request:
This PR adds support for provisioning Windows nodes using vSphere templates in the Rancher2 provider. It enables users to create and manage Windows-based nodes using the same vSphere template logic as for Linux nodes.

## Changes
- Modified the Rancher2 provider to support Windows machine pool in cluster_v2
- Updated the provider logic to handle vSphere Windows template.
- Added new string into tests to ensure proper functionality for Windows nodes.

## Related Issues
- Closes [https://github.com/rancher/terraform-provider-rancher2/issues/1193]
- Closes [https://github.com/rancher/terraform-provider-rancher2/issues/1090]

adonskoy commented:
@matttrach @jiaqiluo can someone review this? This is really lacking feature in the provider

matttrach commented:
@Ihor-kolomiiets thank you so much for your contribution!!
This looks good from the Terraform side of things, but I am getting confirmation from the other teams to make sure it makes sense from the Rancher API and vSphere side of things. In the mean time would you mind signing your commits?
We require all commits signed by a verified key. If we don't hear from you we will probably need to resubmit this under a different account, just letting you know.

HarrisonWAffel commented on docs/resources/node_template.md:
```diff
@@ -436,6 +436,7 @@ The following attributes are exported:
 * `network` - (Optional) vSphere network where the docker VM will be attached (list)
 * `password` - (Optional/Sensitive) vSphere password. Mandatory on Rancher v2.0.x and v2.1.x. Use `rancher2_cloud_credential` from Rancher v2.2.x (string)
 * `pool` - (Optional) vSphere resource pool for docker VM (string)
+* `os` - (Optional) Type of virtual machine OS in vSphere. Default `linux`(string)
```
Are you sure we need to update the node template resource? Isn't that RKE1 specific? Currently we only support rke2 Windows workers, rke1 support has been EOL for a few years

Ihor-kolomiiets replied:
cluster and machine_config in needed, i think i just added it here as well so every resource would have ability to set OS to Windows

HarrisonWAffel reviewed (commented):
generally lgtm, thanks for the contribution! I'm going to test this locally on Monday and give a final review at that point. Only a single question for the time being

Ihor-kolomiiets committed:
add possibility of installing windows worker node with vSphere

Ihor-kolomiiets commented:
> @Ihor-kolomiiets thank you so much for your contribution!! This looks good from the Terraform side of things, but I am getting confirmation from the other teams to make sure it makes sense from the Rancher API and vSphere side of things. In the mean time would you mind signing your commits? We require all commits signed by a verified key. If we don't hear from you we will probably need to resubmit this under a different account, just letting you know.

i resubmited my commit under one name, with GPG key, because one of the commit was with working mail

matttrach commented:
@Ihor-kolomiiets please update your branch with the latest changes from master?

HarrisonWAffel commented:
@Ihor-kolomiiets  thanks for raising this PR1

Also, thanks for your patience while I test this PR locally. I was able to use these changes to provision a Windows vSphere node driver cluster in our test environment. I think this LGTM, but I have two comments

1. It seems the unit tests added in this PR are failing, could you take a look? 
2. I'm still not convinced that we need to add the `os` field to the node template resource, as that resource is specific to rke1 and rke1 Windows nodes have not been supported in over a year. This isn't a hill I'm willing to die on though, and since rke1 is EOL in rancher 2.12 and we should probably clean up a few of the rke1 resources in the provider anyway.

Ihor-kolomiiets committed:
fix machine config tests related to machine os

Ihor-kolomiiets committed:
Merge branch 'rancher:master' into master

Ihor-kolomiiets commented:
> @Ihor-kolomiiets thanks for raising this PR1
> 
> Also, thanks for your patience while I test this PR locally. I was able to use these changes to provision a Windows vSphere node driver cluster in our test environment. I think this LGTM, but I have two comments
> 
> 1. It seems the unit tests added in this PR are failing, could you take a look?
> 2. I'm still not convinced that we need to add the `os` field to the node template resource, as that resource is specific to rke1 and rke1 Windows nodes have not been supported in over a year. This isn't a hill I'm willing to die on though, and since rke1 is EOL in rancher 2.12 and we should probably clean up a few of the rke1 resources in the provider anyway.

I have updated tests, and locally it passes for me, can't run it here to test

also i removed `os`  field on node template, since it's EOL

matttrach committed:
Merge branch 'master' into master
